nucypher/blockchain/eth/actors.py

```diff
@@ -1194,15 +1198,80 @@ def __init__(self,
 
             self.work_tracker = work_tracker or WorkTracker(worker=self)
             if start_working_now:
                 self.work_tracker.start(act_now=False)
 
+    @classmethod
+    def block_until_ready(cls,
+                          worker_address: str,
+                          registry: 'BaseContractRegistry',
+                          poll_rate: int = None,
+                          timeout: int = None) -> str:
+        """
+        Polls the staking_agent and blocks until the staking address is not
+        a null address for the given worker_address.
+        Once the worker is bonded, it returns the staker address.
+        """
+
+        timeout = timeout or cls.BONDING_TIMEOUT
+        poll_rate = poll_rate or cls.BONDING_POLL_RATE
+        staking_agent = ContractAgency.get_agent(StakingEscrowAgent, registry=registry)
+        client = staking_agent.blockchain.client
+        start = maya.now()
+        emitter = StdoutEmitter()  # TODO: Make injectable, or embed this logic into Ursula
+
+        counter, funded, bonded = 0, False, False
+        while True:
+
+            # Read
+            staking_address = staking_agent.get_staker_from_worker(worker_address)
+            ether_balance = client.get_balance(worker_address)
+
+            # Bonding
+            if staking_address == BlockchainInterface.NULL_ADDRESS:
+                if counter % 10 == 0:
+                    emitter.message("Waiting for bonding...", color="yellow")
+            else:
+                if not bonded:
+                    bonded = True
+                    emitter.message(f"Worker is bonded to ({staking_address})!", color='green', bold=True)
+
+            # Balance
+            if not ether_balance:
+                if counter % 10 == 0:
+                    emitter.message("Waiting for ETH funding...", color="yellow")
+            else:
+                if not funded:
+                    funded = True
+                    emitter.message(f"Worker is funded with {Web3.fromWei(ether_balance, 'ether')} ETH)!", color='green', bold=True)
+
+            # Success and Escape
+            if staking_address != BlockchainInterface.NULL_ADDRESS and ether_balance:
+                emitter.message(f"Starting services...", color='yellow', bold=True)
+                break
+
+            # Crash
+            if timeout:
+                now = maya.now()
+                delta = now - start
+                if delta.total_seconds() >= timeout:
+                    if staking_address == BlockchainInterface.NULL_ADDRESS:
+                        raise cls.DetachedWorker(f"Worker {worker_address} not bonded after waiting {timeout} seconds.")
+                    elif not ether_balance:
+                        raise RuntimeError(f"Worker {worker_address} has no ether after waiting {timeout} seconds.")
+
+            # Increment
+            time.sleep(poll_rate)
```

Comment:
I'm not sure how I feel about all these conditionals occurring in this classmethod. It's just a lot going on here. Other than that, it's fine.

Reply:
:heavy_check_mark: